### PR TITLE
network: takes more route information into hash func

### LINKED ISSUE #13506: duplicate routes after customize [Route].
**systemd version the issue has been seen with**
> from the beginning, currently 243.

**Used distribution**
 > Archlinux

**Expected behaviour you didn't see**
 >
```
~ $ ip route
default via 10.0.0.1 dev eth0 proto static initcwnd 88 initrwnd 88 quickack 1 fastopen_no_cookie 1
```

**Unexpected behaviour you saw**
 >
```
~ $ ip route
default via 10.0.0.1 dev eth0 proto static initcwnd 88 initrwnd 88 quickack 1 fastopen_no_cookie 1
default via 10.0.0.1 dev eth0 proto static
```

**Steps to reproduce the problem**
```
/etc/systemd/network $ cat 20-default.network
[Match]
Name=eth0

[Network]
Address=10.0.0.4/24
DHCP=no
DNS=8.8.8.8
Gateway=10.0.0.1
IPv6AcceptRA=no
LinkLocalAddressing=no
LLMNR=false
LLDP=false

[Route]
FastOpenNoCookie=true
Gateway=10.0.0.1
InitialAdvertisedReceiveWindow=88
InitialCongestionWindow=88
QuickAck=true
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #13506.

cc @ssahani and @filbranden.

### CODE REVIEW

File: src/network/networkd-route.c [hunk 1]
```diff
@@ -140,8 +140,10 @@ void route_free(Route *route) {
         network_config_section_free(route->section);
 
         if (route->link) {
-                set_remove(route->link->routes, route);
-                set_remove(route->link->routes_foreign, route);
+                if (set_get(route->link->routes, route) == route)
+                        set_remove(route->link->routes, route);
```
Comment (ssahani) [inline src/network/networkd-route.c]: Sorry I am not able to follow this patch if there is no route it will not be removed so why we checking ?

**Comment by keszybz** [comment]
Is the needs-rework label valid, or should this be reviewed?

**Comment by yuwata** [comment]
This is still in WIP. So, `dont-merge` tag is appropriate.

File: src/network/networkd-route.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-route.c]: Let's consider the case that there exists two `Route` object `a` and `b` such that they are equivalent in the base of `route_hash_func()`, and `a` is stored in the `Set`s. When `b` is freed, then the entry of `a` is removed from the `Set`s, thus `a` is leaked.

Of course, this is a 'potential' memory leak at this time, and the above path did not occurs. This is just a preparation for the following commits.

**Comment by yuwata** [comment]
@keszybz, @ssahani and @filbranden Now all CIs are green. PTAL.

File: src/network/networkd-route.c (hunk 1 above)
Comment (keszybz) [reply src/network/networkd-route.c]: Hmm, if two items in the set appear as equal, but they are not really equivalent, then a set should not be used. When we look up the key in the set, `hash_ops->compare()` is called at some point. If that function returns true, even though it shouldn't, then it should be updated to take more information into account.

The other patches in the series add more stuff to the hash function. Maybe `route_compare_func()` needs an update too?

Comment (yuwata) [reply src/network/networkd-route.c]: @keszybz and @ssahani After re-considering the code, I conclude that this change is not necessary. Hence, dropped now.

**Comment by yuwata** [comment]
@keszybz and @ssahani PTAL. The change is only dropping the first commit.

**Comment by keszybz** [comment]
So... after reviewing this twice, I finally grok what makes the difference: it's not the hash function changes which matter (because hash precision only matters for efficiency, and should not make user visible results). It's the _compare() function changes which fix the bug.

I'll follow up with a cleanup PR which makes all _hash() and _compare() functions in networkd use the same order.

**Pull request merged**